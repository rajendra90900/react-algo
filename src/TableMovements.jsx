import React, { useState, useEffect } from 'react';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import EditIcon from '@mui/icons-material/Edit';
import transactions from '../src/data.json';
import { Button, TextField, Checkbox } from '@mui/material';

export const TableMovement = () => {
    const [transactionsData, setTransactions] = useState(transactions);
    const updateValue = (name, value, id) => {
        let data;
        if (name === 'email') {
            data = transactionsData.map(obj =>
                obj.id === id ? { ...obj, "email": value.target.value } : obj
            );
        };
        setTransactions(data);
    };
    const handleChange = (e, i, id) => {
        const data = transactionsData.map(obj =>
            obj.id === id ? { ...obj, "complete": i !== -1, "isEditable": i } : obj
        );
        setTransactions(data);
    };

    const deleteRow = (id, index) => {
        console.log(transactionsData.filter(object => {
            return object.id !== id;
        }));
        console.log('1111', transactionsData);
        setTransactions(transactionsData.filter(object => {
            return object.id !== id;
        }));
    };

    // useEffect(() => {
    //     // Update the document title using the browser API
    //     setTransactions(transactionsData);
    //   });
    const addRow = () => {
        console.log('123', transactionsData);
        const data = {
            "id": 1,
            "name": "deere",
            "email": "Sat@mail.com",
            "adderss": "South street",
            "transaction_date": "2022-01-02T00:00:00.000Z",
            "amount": 80,
            "points": 0,
            "isEditable": false,
            "complete": false
        };
        transactionsData.push(data)
    }
    return (
        <React.Fragment>
            <h2>Transaction Table</h2>
            <TableContainer component={Paper} style={{ padding: 20 }}>
                <Table sx={{ minWidth: 650 }} aria-label="simple table" sx={{ border: 1 }}>
                    <TableHead sx={{ border: 1 }}>
                        <TableRow sx={{ border: 1 }}>
                            <TableCell align="right" sx={{ border: 1 }}>Id</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Date</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Name</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Email</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Delete</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Edit</TableCell>
                            <TableCell align="right" sx={{ border: 1 }}>Edit</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {transactionsData.map((row, index) => (
                            <TableRow
                                key={row.id}
                                sx={{ border: 2 }}
                            >
                                <TableCell component="th" scope="row">
                                    {row.id}
                                </TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>{row.name}</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>{row.transaction_date}</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>
                                    <TextField
                                        id="standard-basic"
                                        label="Standard"
                                        variant="standard"
                                        value={row.email}
                                        disabled={row.isEditable}
                                        onChange={(e) => updateValue('email', e, row.id)}
                                    />
                                </TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>
                                    <Button
                                        onClick={() => deleteRow(row.id, index)}
                                        disabled={row.isEditable}
                                    >
                                        <DeleteForeverIcon />
                                    </Button>
                                </TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>
                                    <Button disabled={row.isEditable} onClick={() => alert()}><EditIcon /></Button>
                                </TableCell>
                                <TableCell>
                                    <Checkbox
                                        checked={row.checked}
                                        onChange={(e, i) => handleChange(e, i, row.id)}
                                    />
                                </TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
            <Button onClick ={() => addRow()}>Add Row</Button>
        </React.Fragment>
    );
};