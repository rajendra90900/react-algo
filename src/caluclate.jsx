import React from 'react';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import TextField from '@mui/material/TextField';
import Chip from '@mui/material/Chip';
import Stack from '@mui/material/Stack';
import transactions from '../src/data.json';

export class CaluclateAmount extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            totalAmount: 0,
            points: 0,
            transactionsData: transactions,
        };
    };

    componentDidMount() {
        this.totalAmount();
    }

    totalAmount = () => {
        const sum = this.state.transactionsData.reduce((acc, dataObj) => {
            return acc + dataObj.amount;
        }, 0);
        this.caluclatePoints(sum);
        const data = this.state.transactionsData.map(obj =>
            obj.id ? { ...obj, points: this.caluclatePoints(obj.amount) } : obj
        );
        this.setState({ totalAmount: sum, transactionsData: data }, () => this.caluclatePoints(sum));
    };

    caluclatePoints = (amount) => {
        let points;
        if (amount === 0 || amount <= 50) {
            points = 0;
            this.setState({ points });
            return 0;
        };
        if (amount >= 50 && amount <= 100) {
            points = amount - 50;
            this.setState({ points });
            return amount - 50;
        }
        if (amount > 100) {
            points = 50 + ((amount - 100) * 2);
            this.setState({ points });
            return points;
        }
        return;
    };

    caluclateAmountChange = (e, index, id) => {
        const updatedData = this.state.transactionsData.map(obj =>
            obj.id === id ? { ...obj, amount: parseInt(e.target.value) } : obj
        );
        this.setState({ transactionsData: updatedData }, () => this.totalAmount());
        this.totalAmount();
    };

    dateSelected = (month) => {
        let result;
        if (month) {
            const ed = new Date(`2022-${month}-31T00:00:00.000Z`).getTime();
            const sd = new Date(`2022-${month}-01T00:00:00.000Z`).getTime();
            result = transactions.filter(d => {
                var time = new Date(d.transaction_date).getTime();
                return (sd < time && time < ed);
            });

        } else {
            const ed = new Date(`2022-05-31T00:00:00.000Z`).getTime();
            const sd = new Date(`1900-05-01T00:00:00.000Z`).getTime();
            result = transactions.filter(d => {
                var time = new Date(d.transaction_date).getTime();
                return (sd < time && time < ed);
            });
        }
        this.setState({ transactionsData: result }, () => this.totalAmount());
    };
    render() {
        const { points } = this.state;
        return (
            <React.Fragment>
                <h2>Transaction Table</h2>
                <Stack direction="row" spacing={1} style={{ padding: 20 }}>
                    <Chip label="January" onClick={() => this.dateSelected('01')} />
                    <Chip label="Febuary" onClick={() => this.dateSelected('02')} />
                    <Chip label="March" onClick={() => this.dateSelected('03')} />
                    <Chip label="All Time" onClick={() => this.dateSelected()} />

                </Stack>
                <TableContainer component={Paper} style={{ padding: 20 }}>
                    <Table sx={{ minWidth: 650 }} aria-label="simple table" sx={{ border: 1 }}>
                        <TableHead sx={{ border: 1 }}>
                            <TableRow sx={{ border: 1 }}>
                                <TableCell align="right" sx={{ border: 1 }}>Id</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>Date</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>Name</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>Email</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>Amount</TableCell>
                                <TableCell align="right" sx={{ border: 1 }}>Points</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {this.state.transactionsData.map((row, index) => (
                                <TableRow
                                    key={row.id}
                                    sx={{ border: 2 }}
                                >
                                    <TableCell component="th" scope="row">
                                        {row.id}
                                    </TableCell>
                                    <TableCell align="right" sx={{ border: 1 }}>{row.name}</TableCell>
                                    <TableCell align="right" sx={{ border: 1 }}>{row.transaction_date}</TableCell>
                                    <TableCell align="right" sx={{ border: 1 }}>{row.email}</TableCell>
                                    <TableCell align="right" sx={{ border: 1 }}>
                                        <TextField value={row.amount || 0} onChange={(value) => this.caluclateAmountChange(value, index, row.id)} />
                                    </TableCell>
                                    <TableCell align="right" sx={{ border: 1 }}>{row.points}</TableCell>
                                </TableRow>
                            ))}
                        </TableBody>
                    </Table>
                </TableContainer>
                {/* <h3>Total Amount</h3> */}
                <h2 align="right">Total Points: {points}</h2>
            </React.Fragment>
        );
    }
};