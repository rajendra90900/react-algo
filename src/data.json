[
    {
        "id": 1,
        "name": "deere",
        "email": "Sat@mail.com",
        "adderss": "South street",
        "transaction_date": "2022-01-02T00:00:00.000Z",
        "amount": 80,
        "points": 0
    },
    {
        "id": 2,
        "name": "John",
        "email": "john@mail.com",
        "adderss": "North street",
        "transaction_date": "2022-01-13T00:00:00.000Z",
        "amount": 30,
        "points": 0
    },
    {
        "id": 3,
        "name": "Robert",
        "email": "robert@mail.com",
        "adderss": "East street",
        "transaction_date": "2022-01-21T00:00:00.000Z",
        "amount": 55,
        "points": 0
    },
    {
        "id": 4,
        "name": "Mani",
        "email": "mani@mail.com",
        "adderss": "West street",
        "transaction_date": "2022-02-21T00:00:00.000Z",
        "amount": 0,
        "points": 0
    },
    {
        "id": 5,
        "name": "pavi",
        "email": "Sat@mail.com",
        "adderss": "South street",
        "transaction_date": "2022-02-01T00:00:00.000Z",
        "amount": 45,
        "points": 0
    },
    {
        "id": 6,
        "name": "ram",
        "email": "john@mail.com",
        "adderss": "North street",
        "transaction_date": "2022-02-14T00:00:00.000Z",
        "amount": 56,
        "points": 0
    },
    {
        "id": 7,
        "name": "david",
        "email": "robert@mail.com",
        "adderss": "East street",
        "transaction_date": "2022-02-22T00:00:00.000Z",
        "amount": 71,
        "points": 0
    },
    {
        "id": 8,
        "name": "jason",
        "email": "Sat@mail.com",
        "adderss": "South street",
        "transaction_date": "2022-03-02T00:00:00.000Z",
        "amount": 92,
        "points": 0
    },
    {
        "id": 9,
        "name": "holder",
        "email": "john@mail.com",
        "adderss": "North street",
        "transaction_date": "2022-03-16T00:00:00.000Z",
        "amount": 2,
        "points": 0
    },
    {
        "id": 10,
        "name": "roy",
        "email": "robert@mail.com",
        "adderss": "East street",
        "transaction_date": "2022-03-26T00:00:00.000Z",
        "amount": 9,
        "points": 0
    }
]