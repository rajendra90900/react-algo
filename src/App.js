import React from 'react';
import { CaluclateAmount } from './caluclate';

function App() {
  return (
    <div>
      <CaluclateAmount />
    </div>
  );
}

export default App;
