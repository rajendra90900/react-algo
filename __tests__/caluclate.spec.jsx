import React from 'react';
import { render } from '@testing-library/react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

configure({adapter: new Adapter()});

import { CaluclateAmount } from '../src/caluclate';

test('renders caluclate comp', () => {
  const snap = shallow(<CaluclateAmount />);
  expect(snap).toMatchSnapshot();
});
