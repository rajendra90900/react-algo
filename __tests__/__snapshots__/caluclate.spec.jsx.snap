// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders caluclate comp 1`] = `ShallowWrapper {}`;
